File: ts/lcof/50.ts
// 哈希表
function firstUniqChar(s: string): string {
  const ht = {}, len = s.length
  let idx = len
  for (let i = 0; i < len; i++) {
    if (ht[s[i]]) {
      ht[s[i]][0] += 1
    } else {
      ht[s[i]] = [1, i]
    }
  }

  Object.keys(ht).forEach(key => {
    if (ht[key][0] === 1) {
      idx = Math.min(idx, ht[key][1])
    }
  })


  return idx === len ? ' ' : s[idx]
};

File: py3/lcof/50.py
class Solution:
    def firstUniqChar(self, s: str) -> str:
        dic, s_len = {}, len(s)
        for i, v in enumerate(s):
            if v in dic:
                dic[v] = -1
            else:
                dic[v] = i
        idx = s_len
        for v in dic.values():
            if v > -1 and v < idx:
                idx = v
        return s[idx] if idx < s_len else ' '